### PR TITLE
ddl: optimize adding multiple indexes

### LINKED ISSUE #41602: Optimize multi-schema change for `add indexes`
## Enhancement
We can combine multiple "add index" operations in a multi-schema change, which will reduce table data scans and accelerate adding indexes speed.

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #41602

Problem Summary:

### What is changed and how it works?

This PR implements an optimization when adding multiple indexes.

Previously, TiDB scans table records multiple times.  After this PR, TiDB scans only once and extracts all the needed columns. Compared with previous implementation, coprocessor requests cound be greatly reduced.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Benjamin2037 — APPROVED** [review]
LGTM

File: ddl/index_cop.go [hunk 1]
```diff
@@ -435,23 +280,27 @@ func (c *CopContext) buildTableScan(ctx context.Context, startTS uint64, start,
...
 
-func (c *CopContext) fetchTableScanResult(ctx context.Context, result distsql.SelectResult,
-	chk *chunk.Chunk) (bool, error) {
+func fetchTableScanResult(
+	ctx context.Context,
+	copCtx copr.CopContext,
+	result distsql.SelectResult,
+	chk *chunk.Chunk,
+) (bool, error) {
+	c := copCtx.GetBase()
```
Comment (okJiang) [inline ddl/index_cop.go:292]: how about this?
```suggestion
func fetchTableScanResult(
	ctx context.Context,
	copCtx *copr.CopContextBase,
	result distsql.SelectResult,
	chk *chunk.Chunk,
) (bool, error) {
```

File: ddl/index_cop.go [hunk 2]
```diff
@@ -256,165 +257,9 @@ func (c *copReqSenderPool) recycleChunk(chk *chunk.Chunk) {
...
-			fts = append(fts, col.GetType())
-		}
-	}
-	return offsets, fts
-}
-
-func (c *CopContext) buildTableScan(ctx context.Context, startTS uint64, start, end kv.Key) (distsql.SelectResult, error) {
-	dagPB, err := buildDAGPB(c.sessCtx, c.tblInfo, c.colInfos)
+func buildTableScan(ctx context.Context, copCtx copr.CopContext, startTS uint64, start, end kv.Key) (distsql.SelectResult, error) {
+	c := copCtx.GetBase()
```
Comment (okJiang) [inline ddl/index_cop.go:261]: ditto

File: ddl/index.go [hunk 3]
```diff
@@ -637,35 +643,53 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
...
 		// write only -> reorganization
-		indexInfo.State = model.StateWriteReorganization
-		_, err = checkPrimaryKeyNotNull(d, w, t, job, tblInfo, indexInfo)
-		if err != nil {
+		var checkErr bool
+		for _, indexInfo := range allIndexInfos {
+			indexInfo.State = model.StateWriteReorganization
+			_, err = checkPrimaryKeyNotNull(d, w, t, job, tblInfo, indexInfo)
+			if err != nil {
+				checkErr = true
```
Comment (okJiang) [inline ddl/index.go:685]: ditto

File: ddl/index.go [hunk 4]
```diff
@@ -1074,40 +1120,50 @@ func onDropIndex(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
 			return ver, errors.Trace(err)
 		}
 	case model.StateWriteOnly:
 		// write only -> delete only
-		indexInfo.State = model.StateDeleteOnly
-		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, originalState != indexInfo.State)
+		for _, indexInfo := range allIndexInfos {
+			indexInfo.State = model.StateDeleteOnly
+		}
+		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, originalState != allIndexInfos[0].State)
```
Comment (okJiang) [inline ddl/index.go:1143]: ditto

File: ddl/index.go [hunk 5]
```diff
@@ -1074,40 +1120,50 @@ func onDropIndex(d *ddlCtx, t *meta.Meta, job *model.Job) (ver int64, _ error) {
...
 			return ver, errors.Trace(err)
 		}
 	case model.StateDeleteOnly:
 		// delete only -> reorganization
-		indexInfo.State = model.StateDeleteReorganization
-		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, originalState != indexInfo.State)
+		for _, indexInfo := range allIndexInfos {
+			indexInfo.State = model.StateDeleteReorganization
+		}
+		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, originalState != allIndexInfos[0].State)
```
Comment (okJiang) [inline ddl/index.go:1152]: ditto

File: ddl/ingest/config.go [hunk 6]
```diff
@@ -55,13 +55,9 @@ func genConfig(ctx context.Context, memRoot MemRoot, jobID int64, unique bool, i
...
 	adjustImportMemory(ctx, memRoot, cfg)
 	cfg.Checkpoint.Enable = true
-	if unique {
-		cfg.TikvImporter.DuplicateResolution = lightning.DupeResAlgErr
-		// TODO(lance6716): will introduce fail-fast for DDL usage later
-		cfg.Conflict.Threshold = math.MaxInt64
-	} else {
-		cfg.TikvImporter.DuplicateResolution = lightning.DupeResAlgNone
-	}
+	cfg.TikvImporter.DuplicateResolution = lightning.DupeResAlgErr
```
[the remark this answers is not in the document]
Comment (tangenta) [reply ddl/ingest/config.go:61]: Updated.

File: ddl/index.go [hunk 7]
```diff
@@ -637,35 +643,53 @@ func (w *worker) onCreateIndex(d *ddlCtx, t *meta.Meta, job *model.Job, isPK boo
...
 		// delete only -> write only
-		indexInfo.State = model.StateWriteOnly
-		_, err = checkPrimaryKeyNotNull(d, w, t, job, tblInfo, indexInfo)
-		if err != nil {
+		var checkErr bool
+		for _, indexInfo := range allIndexInfos {
+			indexInfo.State = model.StateWriteOnly
+			_, err = checkPrimaryKeyNotNull(d, w, t, job, tblInfo, indexInfo)
+			if err != nil {
+				checkErr = true
```
[the remark this answers is not in the document]
Comment (tangenta) [reply ddl/index.go:666]: Do you mean use label instead?

```go
SWITCH_STATE:
  switch {
    case X:
      for {
        //...
        break SWITCH_STATE
      }
  }
```

**Review by okJiang — COMMENTED** [review]
rest lgtm

File: ddl/multi_schema_change.go [hunk 8]
```diff
@@ -49,6 +49,7 @@ func (d *ddl) MultiSchemaChange(ctx sessionctx.Context, ti ast.Ident) error {
 	if err != nil {
 		return errors.Trace(err)
 	}
+	mergeAddIndex(ctx.GetSessionVars().StmtCtx.MultiSchemaInfo)
```
[the remark this answers is not in the document]
Comment (tangenta) [reply ddl/multi_schema_change.go:52]: Then we need to allocate another slice to store the result..

**Review by okJiang — APPROVED** [review]

**Review by zanmato1984 — APPROVED** [review]

**Comment by okJiang** [comment]
need fix conflict @tangenta @Benjamin2037

**Pull request merged**